File: LeetCode/1475. Final Prices With a Special Discount in a Shop/solution.py
"""
103 / 103 test cases passed.
Runtime: 40 ms
Memory Usage: 14.9 MB
"""
class Solution:
    def finalPrices(self, prices: List[int]) -> List[int]:
        for i, pricei in enumerate(prices):
           for j, pricej in enumerate(prices[i + 1:]):
               if pricej <= pricei:
                   prices[i] = pricei - pricej 
                   break
        return prices

File: LeetCode/1475. Final Prices With a Special Discount in a Shop/solution.cpp
/**
 * 103 / 103 test cases passed.
 * Runtime: 4 ms
 * Memory Usage: 9.6 MB 
 */
class Solution {
public:
    vector<int> finalPrices(vector<int>& prices) {
        int n = prices.size();
        vector<int> ans(n);
        stack<int> s;
        for (int j = 0; j < n; ++ j) {
            while (!s.empty() && prices[j] <= prices[s.top()]) {
                int i = s.top(); s.pop();
                ans[i] = prices[i] - prices[j];
            }
            s.push(j);
            ans[j] = prices[j];
        }
        return ans;
    }
};
